File: csharpcore/GildedRose/GildedRose.cs
using System;
using System.Collections.Generic;

namespace GildedRoseKata
{
    public class GildedRose
    {
        private const string AgedBrie = "Aged Brie";
        private const string BackstagePasses = "Backstage passes to a TAFKAL80ETC concert";
        private const string Sulfuras = "Sulfuras, Hand of Ragnaros";

        IList<Item> Items;

        public GildedRose(IList<Item> Items)
        {
            this.Items = Items;
        }

        public void AdjustQuality(Item item, int delta)
        {
            item.Quality += delta;
            item.Quality = Math.Clamp(item.Quality, 0, 50);
        }

        public void UpdateQuality()
        {
            foreach (var item in Items)
            {
                if (item.Name == Sulfuras)
                {
                    continue;
                }

                item.SellIn -= 1;

                var delta = item.Name switch
                {
                    AgedBrie => item.SellIn < 0 ? 2 : 1,
                    BackstagePasses => item.SellIn switch
                    {
                        var i when i < 0 => -item.Quality,
                        var i when i < 5 => 3,
                        var i when i < 10 => 2,
                        _ => 1
                    },
                    var name when name.StartsWith("Conjured") => item.SellIn < 0 ? -4 : -2,
                    _ => item.SellIn < 0 ? -2 : -1
                };

                AdjustQuality(item, delta);
            }
        }
    }
}

File: go/gilded-rose.go
package main

import "strings"

const (
	AgedBrie        = "Aged Brie"
	BackstagePasses = "Backstage passes to a TAFKAL80ETC concert"
	Sulfuras        = "Sulfuras, Hand of Ragnaros"
	ConjuredPrefix  = "Conjured"
	MaxQuality      = 50
	MinQuality      = 0
)

type Item struct {
	name            string
	sellIn, quality int
}

func clamp(i int) int {
	if i < MinQuality {
		return MinQuality
	}
	if i > MaxQuality {
		return MaxQuality
	}
	return i
}

func UpdateQuality(items []*Item) {
	for _, item := range items {

		if item.name == Sulfuras {
			continue
		}

		switch {
		case item.name == AgedBrie:
			item.quality = clamp(item.quality + 1)
		case item.name == BackstagePasses:
			item.quality = clamp(item.quality + 1)

			if item.sellIn < 11 {
				item.quality = clamp(item.quality + 1)
			}
			if item.sellIn < 6 {
				item.quality = clamp(item.quality + 1)
			}
		case strings.HasPrefix(item.name, ConjuredPrefix):
			item.quality = clamp(item.quality - 2)
		default:
			item.quality = clamp(item.quality - 1)
		}

		item.sellIn = item.sellIn - 1

		if item.sellIn < 0 {
			switch {
			case item.name == AgedBrie:
				item.quality = clamp(item.quality + 1)
			case item.name == BackstagePasses:
				item.quality = clamp(item.quality - item.quality)
			case strings.HasPrefix(item.name, ConjuredPrefix):
				item.quality = clamp(item.quality - 2)
			default:
				item.quality = clamp(item.quality - 1)
			}
		}
	}
}
